ESQL: Improve field reference tracking in `FORK` command

kanoshiou opened the pull request:
This change fixes the field resolution inefficiency in `FORK` commands where queries like 
```
FROM employees | fork (eval x = 1 | keep x) (eval y = 2 | keep y) (eval z = 3 | keep z) 
```
were unnecessarily requesting all fields (*) from `field_caps` instead of only the minimal required fields. 

The fix introduces branch-specific `KEEP` reference tracking through `currentBranchKeepRefs` and replaces the overly conservative empty reference check with nuanced logic that properly distinguishes between branches that genuinely need all fields (like filter-only branches) versus those with explicit field constraints, resulting in more efficient field resolution and reduced overhead during the logical plan optimization step.

Closes #137283

kanoshiou committed:
refactor(esql): Improve field reference tracking in FORK command

- Enhance field reference tracking for FORK command branches
- Add branch-specific keep references tracking with `currentBranchKeepRefs`
- Refine logic for determining when to project all fields in FORK branches
- Update field collection strategy to handle more complex query scenarios
- Modify test cases to validate new field reference tracking behavior
The changes improve the field reference collection mechanism in ESQL's FORK command, providing more precise field selection and projection logic across different branch scenarios.

kanoshiou committed:
Update tests

kanoshiou committed:
Update docs/changelog/137678.yaml

kanoshiou committed:
Merge branch 'main' into fork-branch-keep-refs-tracking

elasticsearchmachine commented:
Pinging @elastic/es-search-relevance (Team:Search Relevance)

fang-xing-esql reviewed (commented):
Thanks for the PR, @kanoshiou! Looking at the changes in the existing tests, I noticed some `FORK` tests changed to `ALL_FIELDS`. Is there a specific reason why they need to be `ALL_FIELDS`? The old(existing) behavior of `FORK` looks fine to me, but let me know if I'm missing something.

The original issue was created for subqueries, so I didn't expect the behavior changes to `FORK` unless there is a good reason.

Also a heaps up to @ioanatia, to confirm whether the changes related to `FORK` are fine.

ioanatia reviewed (commented):
thank you for looking into this.
while I agree this fixes a bug we currently have, @fang-xing-esql correctly noted that we might be returning more fields than we need
I left a suggestion for a change.
In the long term, I think FieldNameUtils would benefit from a refactor, as I myself find it hard to follow up what is going on.

kanoshiou committed:
Merge branch 'refs/heads/main' into fork-branch-keep-refs-tracking

kanoshiou committed:
Consider column-reducing plans after FORK when resolving field names

Add a `reduceColumnsAfterFork` flag to track if there are plans after
FORK that reduce columns to a known set (e.g., Project, Aggregate).

Previously, for queries like:
  from test | fork (eval x = 1 | keep x) (where true) | stats c = count(*)

The second branch would trigger projectAll because it
has no KEEP command, causing all fields to be requested from field_caps.

With this change, we now check if there's a column-reducing plan after
FORK before deciding to request all fields. If such a plan exists
(like `stats`), we only request the necessary fields.

Commit:
[CI] Auto commit changes from spotless

ioanatia commented:
@elasticsearchmachine test this please
